firebase-perf/src/test/java/com/google/firebase/perf/application/FrameMetricsRecorderTest.java

```diff
@@ -0,0 +1,405 @@
+// Copyright 2022 Google LLC
```

Comment:
General comment on how this class is being tested: It looks like this class is testing most things through interactions rather than testing behaviors / expected outcomes through public APIs. go/unit-testing-practices#public-apis

I try to write tests from the point of view of a potential caller (i.e. a use-case to use this class). For example, as a caller, I don't care about how this class is implemented - that means that I don't care if a map is called once or twice. What I care about is that the results of my interactions are correct. I'm not sure that the current tests verify that as much as verify implementation details.

To put it differently, these tests are very close to change-detector tests because they are not testing that the class behaves correctly, but rather that the internals of the class have specific interactions.

go/tott/299 has a much better explanation of what I'm trying to say :)

Reply:
That philosophy makes a lot of sense. However I am finding difficulty in achieving those, for example, `start()` and `stop()` are related to each other. `start()` is a `void` function, it does thing so `stop()` will work, but I don't know how to only use outputs to test start individually aside from mocking its implementation detail and seeing behaviors on those mocks (definitely fragile to changes). Similarly `stop()` relies on `start()` in the sense that `stop()` would only work when `start()` has been called prior and correctly, so it'd be hard to test individually without mocking everything in `stop()`.

Reply:
I guess a followup question is: for methods that are dependent on each other or used in conjunction, is it okay to write tests that tests both at the same time instead of independently?

Reply:
Discussed offline. Summary:

Testing the behavior of the class is what we expect is much more useful than testing the internal interactions of the class. The interactions could be what we think they should be, but they could lead to incorrect outcomes. By testing the behavior we can verify that the class does what we want it to do